### ISSUE TITLE
Improve docs on backpressure when forking and re-joining streams

### ISSUE THREAD

**Issue by caolan** [caolan/highland#59]
This issue was spawned by #40

**Comment by dts** [comment]
Bump.  I spent a while trying to figure out how to get a simple example working, I'm going to leave it here in case someone finds my confusion valuable.  I wanted to create a streaming system that would take in a stream of numbers, and prepend and append new numbers to fill out a range.  Desired behavior:
```
fillIn(1,10)([5,6,7]) // put out a stream: [2,3,4,5,6,7,8,9], where 5,6,7 are reused (important if they are objects not integers)
```

My final, working code looks like this:
```js
import _ from 'highland';

var logArray = _.curry(function(name,array) {
  console.log(`${ name } : `,array);
});

function min(a,b) { return a>b?b:a; }
function max(a,b) { return a>b?a:b; }

let numbersBetween = _.curry(function(reference,err,number,push,next) {
  if(err) { push(err); }
  else if (number === _.nil) {
    console.log("Between done, sending on end.");
    push(null, _.nil);
    return;
  } else {
    var n = min(reference,number)+1;
    var end = max(reference,number);
    
    while(n < end) {
      console.log(`pushing ${n}`);
      push(null,n);
      n++;
    }
  }
  
  next();
});

// return a stream that puts out all of the integers between earliest
// and latest, including the ones from an assumed-sorted "existing"
// array.
function getNumbers(lowest,highest,inNumbers) {
  let existing = _(inNumbers || [lowest]);

  let after = existing.observe().take(1).consume(numbersBetween(highest));
  let before = existing.observe().last(1).consume(numbersBetween(lowest));

  before.observe().toArray(logArray('before'));
  existing.observe().toArray(logArray('existing'));
  after.observe().toArray(logArray('after'));
  
  let res = before.concat(after,existing);
  
//  res.toArray(logArray('res'));

  existing.resume();
//  after.resume();
//  before.resume();
    
  return res;
}

getNumbers(1,10,[5,6]).toArray(logArray('FinalRES:'));
```
What took me a long time to figure out is that you have to NOT resume some streams - ("resuming" is a "terminating" event according to some parts of the docs?).  When I don't "resume" `existing`, or DO resume `after` or `before`, everything gets stuck on `concat`, why?  The `logArray` calls from the observed streams are all reporting "finished and did the right thing as expected", but the concatenated stream just doesn't.  I'm guessing that when you don't resume existing it never un-pauses, and when you do resume after and before, they go from being dependent on `existing` to being dead or something.  I don't actually understand at a deep level what 'resumed', 'paused', or otherwise really means.  A few concrete examples with fork/merge/concat/observe, and a more detailed explanation of why you would want to use pipline would go a long way to explaining this situation to me!

**Comment by vqvu** [comment]
Agreed that there needs to be better docs, but I can't promise anything in soon. I can explain a bit more about your specific problem, though.

You can think of streams as being an async generator. Calling `resume` *consumes* from the stream by asking it to start producing data. It's roughly equivalent (though not exactly) to calling `each` with a callback that does nothing. A *paused* stream is not producing data, while a *resumed* or *consumed* stream is. All streams start paused. They only begin producing data when a `consumption` (what you call "terminating") event happens. That is, when a [consumption methods](http://highlandjs.org/#Consumption) or `resume` is called.

When you `observe` a stream, you create a new stream that observes the source stream and remembers what values the source stream produces. It can then replay those values when it (the observing stream) is consumed. Naturally, it can't produce any values if the source stream hasn't produced any value.

Now, `concat` is lazy. When you concat `before` and `after`, it won't consume from `after` until `before` completes (i.e., until `before` produces `nil`). In fact, laziness is a big part of Highland streams, and most other transforms work this way. Now, `before` must wait for `existing` to produce values since it observes `existing`. So then you have a situation where `concat` won't consume from `existing` until `before` produces a `nil`, but `before` won't produce anything until `existing` produces something. But nothing resumes `existing`! You end up with a deadlock arising from the circular dependency. In general, do not introduce circular dependencies into your stream construction and you won't have this problem [1].

As for your solution. I don't think it works. First of all, `concat` only takes one argument, so that call you're making isn't working as expected. It probably won't if `inNumbers` has three or more values. I ran `getNumbers(1,10,[5,6,7])` and got `[ 2, 3, 4, 5, 6, 6, 7, 8, 9 ]`.

Second of all, your solution is overly complicated. In general, it's rare to need to use `resume` directly. You can implement what you want directly with a single `consume`.

```javascript
function pushRange(min, max, push) {
    var i = min + 1;
    while (i < max) {
        push(null, i);
        i++;
    }
}

// Simplifying assumptions:
// - lowest < highest.
// - inNumbers is sorted.
// - inNumbers === [] || lowest < first(inNumbers)
// - inNumbers === [] || highest > last(inNumbers)
function getNumbers(lowest, highest, inNumbers) {
    var nextValue = lowest;
    var existing = _(inNumbers || [lowest]);
    return existing.consume(function (err, x, push, next) {
        if (err != null) {
            push(err);
            next();
        } else if (x === _.nil) {
            // At the end. Push the suffix.
            pushRange(nextValue, highest, push);
            push(null, _.nil);
        } else {
            if (nextValue === lowest) {
                // First time, push the prefix.
                pushRange(nextValue, x, push);
            }
            nextValue = x;
            push(null, x);
            next();
        }
    });
}

getNumbers(1, 10, [4, 5, 6]).toArray(_.log);
// => [ 2, 3, 4, 5, 6, 7, 8, 9 ]
```

[1] If you're interested in how to get circular dependencies to work, see the discussion on flow-based programming in #371.

**Comment by dts** [comment]
Thank you for your response, and you were right, the code I pasted first did not work as expected in general, only for one of the inputs that I happened to use as my initial test case!  Your simplified version makes a lot of sense, and is significantly simpler.  I ended up figuring out many of these details, but if you'll bear with my bad architecture for one moment, I have one final question:

```js
function getNumbers(lowest,highest,inNumbers) {
  let existing = _(inNumbers || [lowest]);

  let after = existing.fork().take(1).consume(numbersBetween(highest));
  let before = existing.fork().last(1).consume(numbersBetween(lowest));

  before.observe().toArray(logArray('before'));
  existing.observe().toArray(logArray('existing'));
  after.observe().toArray(logArray('after'));

  let res = _([before,existing.fork(),after]).sequence();
  existing.resume();
    
  return res;
}
```

In this (now actually working, afaik) example, if I remove the existing.resume(), and the .fork() call on the line above it, when I `.toArray()` the result of this function, it does not work.  My internal explanation for that is that `existing` never got un-paused, though I find this confusing, because shouldn't it get un-paused when `sequence()` begins pulling on it, when I call `toArray()`?

**Comment by vqvu** [comment]
This is a feature of `fork`. Unlike `observe` (which just passively observe its source), forks directly affects their source and share backpressure. This means a few things:

1. No fork of a source will produce any value until *all* forks are consumed.
2. Once all forks all are consumed, the source is automatically consumed to produce values.
3. If any fork is slow, all forks are slowed down to match it.

Here's an example:

```javascript
var source = _([1, 2, 3]);

// fork1 is fast.
var fork1 = source.fork();

// fork2 is slow. It will only consume one item per second.
var fork2 = source.fork()
    .consume(function (err, x, push, next) {
        push(err, x);
        if (x !== _.nil) {
            // Simulate slowness.
            setTimeout(next, 1000);
        }
    });

// Now we consume from fork1.
fork1.each(function (x) {
    console.log('fork1:', x);
});

// Nothing will print yet.

// Only until we consume from fork2 does values start being
// produced, one per second.
fork2.each(function (x) {
    console.log('fork2:', x);
});
```

`before` and `after` both fork `existing`. `sequence` will consume from the source one at a time. That means it will consume from `before` before it even touches `after`. But both `before` and `after` must be consumed before either produce anything, so `sequence` is stuck waiting for `before` to produce data, which is stuck waiting for `after` to be consumed, which is stuck waiting for `sequence` to consume from it.

If you switch `before` and `after` to `observe` and use `let res = _([before, existing.observe(), after]).sequence()`, then you can use `existing.resume()` and everything should work. I don't recommend doing so, however, since you're basically bypassing backpressure.
